src/diameter_handlers.cpp

```diff
@@ -0,0 +1,465 @@
+  irs_vector_success_callback success_cb =
+    std::bind(&RegistrationTerminationTask::get_registration_sets_success, this, std::placeholders::_1);
+
+  failure_callback failure_cb =
+    std::bind(&RegistrationTerminationTask::get_registration_sets_failure, this, std::placeholders::_1);
+
+  // Figure out which registration sets we're de-registering
+  if ((_deregistration_reason != SERVER_CHANGE) &&
+      (_deregistration_reason != NEW_SERVER_ASSIGNED))
+  {
+    // We only record the list of public identities on the request of the reason
+    // isn't SERVER_CHANGE or NEW_SERVER_ASSIGNED (in those cases, we will
+    // deregister all the IRSs for the provided IMPIs)
+    _impus = _rtr.impus();
+  }
+
+  if ((_impus.empty()) && ((_deregistration_reason == PERMANENT_TERMINATION) ||
+                           (_deregistration_reason == REMOVE_SCSCF) ||
+                           (_deregistration_reason == SERVER_CHANGE) ||
+                           (_deregistration_reason == NEW_SERVER_ASSIGNED)))
+  {
+    // If we don't have a list of impus and the reason is allowed, we want to
+    // deregister all of the IRSs for the provided list of IMPIs, so start by
+    // getting all the IRSs for these IMPIs from the cache
+    std::string impis_str = boost::algorithm::join(_impis, ", ");
+    TRC_DEBUG("Looking up registration sets from the cache for IMPIs: %s", impis_str.c_str());
+    SAS::Event event(this->trail(), SASEvent::CACHE_GET_REG_DATA_IMPIS, 0);
+    event.add_var_param(impis_str);
+    SAS::report_event(event);
+
+    _cfg->cache->get_implicit_registration_sets_for_impis(success_cb,
+                                                          failure_cb,
+                                                          _impis,
+                                                          this->trail());
+  }
+  else if ((!_impus.empty()) && ((_deregistration_reason == PERMANENT_TERMINATION) ||
+                                 (_deregistration_reason == REMOVE_SCSCF)))
+  {
+    // If we have a list of IMPUs to deregister and the reason is allowed, we
+    // want to only deregister those specific impus, so start by getting all the
+    // IRSs from the cache for those IMPUS
+    std::string impus_str = boost::algorithm::join(_impus, ", ");
+    TRC_DEBUG("Looking up registration sets from the cache for IMPUs %s", impus_str.c_str());
+    SAS::Event event(this->trail(), SASEvent::CACHE_GET_REG_DATA, 0);
+    event.add_var_param(impus_str);
+    SAS::report_event(event);
+
+    _cfg->cache->get_implicit_registration_sets_for_impus(success_cb,
+                                                          failure_cb,
+                                                          _impus,
+                                                          this->trail());
+  }
+  else
+  {
+    // This is an invalid deregistration reason.
+    TRC_ERROR("Registration-Termination request received with invalid deregistration reason %d",
```

Comment:
Again, existing code (and we don't necessarily have any impus), so I'm tempted to leave as-is